README
NAME
    Barcode::DataMatrix - Generate data for Data Matrix barcodes

SYNOPSIS
        use Barcode::DataMatrix;
        my $data = Barcode::DataMatrix->new->barcode('MONKEY');
        for my $row (@$data) {
            print for map { $_ ? "#" : ' ' } @$row;
            print "\n";
        }

DESCRIPTION
    This class is used to generate data for Data Matrix barcodes. It is
    primarily useful as a data source for barcode modules that do rendering,
    such as HTML::Barcode::DataMatrix. You can easily make a version that
    renders an image, PDF, or anything else.

METHODS
  new (%attributes)
    Instantiate a new Barcode::DataMatrix object. The `%attributes' hash can
    take any of the other attributes listed below.

  barcode ($text)
    Generate barcode data representing the `$text' string. This returns an
    array ref of rows in the data matrix, each containing array refs of
    cells within that row. The cells are true and false values that
    represent filled or empty squares.

    This can throw an exception if it's unable to generate the barcode data.

ATTRIBUTES
  encoding_mode
    The encoding mode for the data matrix. Can be one of: `AUTO' (default),
    `ASCII', `C40', `TEXT', `BASE256', or `NONE'.

  process_tilde
    Set to true to indicate the tilde character "~" is being used to
    recognize special characters. See this page for more information:
    http://www.idautomation.com/datamatrixfaq.html

AUTHOR
    Mons Anderson `<inthrax@gmail.com>' (GD::Barcode::DataMatrix at
    https://github.com/Mons/perl-ex/, from which this distribution
    originates)

    Mark A. Stratman, `<stratman@gmail.com>'

SOURCE REPOSITORY
    http://github.com/mstratman/Barcode-DataMatrix

SEE ALSO
    HTML::Barcode::DataMatrix
    http://grandzebu.net/informatique/codbar-en/datamatrix.htm
    http://www.idautomation.com/datamatrixfaq.html

LICENSE AND COPYRIGHT
    Copyright 2011 the AUTHORs listed above.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Barcode-DataMatrix
0.08    Thu Mar 24 15:23:34 CDT 2016
        - Removed type dependencies

0.07    Fri Jan  1 13:20:21 CST 2016
        - Added missing 'use warnings' pragmas
        - Removed commented-out code
        - Removed now unnecessary "no critic" statement
        - Removed unused constants and duplication
        - Ignore Devel::Cover's cover_db dir
        - Update outdated URLs to new locations

0.06    Wed Dec 30 19:00:10 CST 2015
        - Switched from Any::Moose to Moo
        - Documentation added to the engine
        - Many things cleaned up (see git history)

0.05    Thu Apr 16 09:56:07 CDT 2015
        - Updated the various enums to use an array reference rather
          than an array to avoid deprecation warnings when using Moose 2.1200
          or newer

0.04    Thu Jun 21 09:01:05 CDT 2012
        - Use Module::Install::Repository because metacpan doesn't
          display the repository correctly as set by GithubMeta.
        - Do not auto_include(), and do not include Test::More.

0.03    Mon Jul  4 01:41:08 CDT 2011
        - Move author tests to xt to avoid failed tests.

0.02    Sun Jul  3 02:28:24 CDT 2011
        - Fixed SEE ALSO link

0.01    Sun Jul  3
        - First version, released on an unsuspecting world.



lib/Barcode/DataMatrix.pm
package Barcode::DataMatrix;
use Moo;
use Barcode::DataMatrix::Engine ();

our $VERSION = '0.08';

has 'encoding_mode' => (
    is       => 'ro',
    isa      => sub { my $type = shift; for (qw(ASCII C40 TEXT BASE256 NONE AUTO)) { return 1 if $type eq $_ } return 0; },
    default  => 'AUTO',
    documentation => 'The encoding mode for the data matrix. Can be one of: ASCII C40 TEXT BASE256 NONE AUTO',
);
has 'process_tilde' => (
    is       => 'ro',
    default  => 0,
    documentation => 'Boolean. Set to true to indicate the tilde character "~" is being used to recognize special characters.',
);

=head1 NAME

Barcode::DataMatrix - Generate data for Data Matrix barcodes

=head1 SYNOPSIS

    use Barcode::DataMatrix;
    my $data = Barcode::DataMatrix->new->barcode('MONKEY');
    for my $row (@$data) {
        print for map { $_ ? "#" : ' ' } @$row;
        print "\n";
    }

=cut

=head1 DESCRIPTION

This class is used to generate data for Data Matrix barcodes. It is primarily
useful as a data source for barcode modules that do rendering,
such as L<HTML::Barcode::DataMatrix>.  You can easily make a version that
renders an image, PDF, or anything else.

=head1 METHODS

=head2 new (%attributes)

Instantiate a new Barcode::DataMatrix object. The C<%attributes> hash
can take any of the other L<attributes|/ATTRIBUTES> listed below.

=cut

=head2 barcode ($text)

Generate barcode data representing the C<$text> string.  This returns
an array ref of rows in the data matrix, each containing array refs of
cells within that row. The cells are true and false values
that represent filled or empty squares.

This can throw an exception if it's unable to generate the barcode data.

=cut

sub barcode {
    my ($self, $text) = @_;

    my $engine = Barcode::DataMatrix::Engine->new(
        $text,
        $self->encoding_mode,
        undef, # size
        $self->process_tilde,
    );

    my $rows = $engine->{rows};
    my $cols = $engine->{cols};
    my $bitmap = $engine->{bitmap};
    my $rv = [];
    for my $r (0 .. $rows - 1) {
        my $row = [];
        for my $c (0 .. $cols - 1) {
            push @$row, ($bitmap->[$c]->[$r] ? 1 : 0);
        }
        push @$rv, $row;
    }

    return $rv;
}

=head1 ATTRIBUTES

=head2 encoding_mode

The encoding mode for the data matrix. Can be one of:
C<AUTO> (default), C<ASCII>, C<C40>, C<TEXT>, C<BASE256>, or C<NONE>.

=head2 process_tilde

Boolean. Set to true to indicate the tilde character "~" is being used to recognize
special characters. See this page for more information:
L<http://www.idautomation.com/datamatrixfaq.html>

=cut

=head1 AUTHORS

Mons Anderson C<< <inthrax@gmail.com> >> (GD::Barcode::DataMatrix at L<https://github.com/Mons/perl-ex/>, from which this distribution originates)

Mark A. Stratman, C<< <stratman@gmail.com> >>

Paul Cochrane, L<https://github.com/paultcochrane>

=head1 SOURCE REPOSITORY

L<http://github.com/mstratman/Barcode-DataMatrix>

=head1 SEE ALSO

=over 4

=item L<HTML::Barcode::DataMatrix>

=item L<http://grandzebu.net/informatique/codbar-en/datamatrix.htm>

=item L<http://www.idautomation.com/datamatrixfaq.html>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2015 the AUTHORs listed above.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of Barcode::DataMatrix


lib/Barcode/DataMatrix/Reed.pm
package Barcode::DataMatrix::Reed;

=head1 Barcode::DataMatrix::Reed

This is just a renamed version of Algorithm::DataMatrix::Reed
by Mons Anderson
from http://code.google.com/p/perl-ex/

For a rough explanation of the structure of this code, see
L<https://en.wikiversity.org/wiki/Reed%E2%80%93Solomon_codes_for_coders#Reed.E2.80.93Solomon_codes>.
Note that the link is not the basis of the implementation (this is unknown),
however it helps explain how Reed-Solomon encoding is implemented and hence
how the code below works.

=cut

use strict;
use warnings;
use Carp;

=head2 DEBUG

Control whether or not debugging output is printed.

=cut

sub DEBUG { 0 }

=head2 mult (x, y)

Multiply two Galois field element values and return the result.

=cut

sub mult {
    my ( $x, $y ) = @_;
    return 0 unless $x * $y;
    return $Barcode::DataMatrix::Constants::GFI[
      ( $Barcode::DataMatrix::Constants::GFL[$x] +
            $Barcode::DataMatrix::Constants::GFL[$y] ) % 255
    ];
}

=head2 encode (ai, j)

Encode the message array and return it.

=cut

sub encode {
    my ( $ai, $j ) = @_;
    my $i = @$ai;
    for ( 0 .. $#$ai ) {
        ( $ai->[$_] & 0xFF ) != $ai->[$_]
          and warn("number $ai->[$_] at index $_ is not a byte size"),
          $ai->[$_] = $ai->[$_] & 0xFF;
    }
    warn "CalcReed: ai [@$ai], $j\n" if DEBUG;

    my $p =
      exists $Barcode::DataMatrix::Constants::POLY{$j}
      ? $Barcode::DataMatrix::Constants::POLY{$j}
      : $Barcode::DataMatrix::Constants::POLY{68};
    warn "CalcReed: poly [@$p]\n" if DEBUG;

    @$ai[ $i .. $i + $j - 1 ] = (0) x $j;
    for my $l ( 0 .. $i - 1 ) {
        my $word0 = ( $ai->[$i] ^ $ai->[$l] );
        for my $i1 ( 0 .. $j - 2 ) {
            $ai->[ $i + $i1 ] =
              ( $ai->[ $i + $i1 + 1 ] ^ mult( $word0, $p->[$i1] ) );
        }
        $ai->[ $i + $j - 1 ] = mult( $word0, $p->[ $j - 1 ] );
    }
    warn "CalcReed: result [@$ai]\n" if DEBUG;
    return $ai;
}

1;


lib/Barcode/DataMatrix/Constants.pm
package Barcode::DataMatrix::Constants;

=head1 Barcode::DataMatrix::Constants

Constants associated with the Reed-Solomon error correction encoding used
for data matrices.  See
L<https://en.wikipedia.org/wiki/Reed%E2%80%93Solomon_error_correction> for
more information.

=cut

use strict;
use warnings;

=head2 GFI

Exponent values used in the multiplication step of the Reed-Solomon encoding
algorithm.

=cut

our @GFI = (
    1,   2,   4,   8,   16,  32,  64,  128, 45,  90,  180, 69,  138, 57,
    114, 228, 229, 231, 227, 235, 251, 219, 155, 27,  54,  108, 216, 157,
    23,  46,  92,  184, 93,  186, 89,  178, 73,  146, 9,   18,  36,  72,
    144, 13,  26,  52,  104, 208, 141, 55,  110, 220, 149, 7,   14,  28,
    56,  112, 224, 237, 247, 195, 171, 123, 246, 193, 175, 115, 230, 225,
    239, 243, 203, 187, 91,  182, 65,  130, 41,  82,  164, 101, 202, 185,
    95,  190, 81,  162, 105, 210, 137, 63,  126, 252, 213, 135, 35,  70,
    140, 53,  106, 212, 133, 39,  78,  156, 21,  42,  84,  168, 125, 250,
    217, 159, 19,  38,  76,  152, 29,  58,  116, 232, 253, 215, 131, 43,
    86,  172, 117, 234, 249, 223, 147, 11,  22,  44,  88,  176, 77,  154,
    25,  50,  100, 200, 189, 87,  174, 113, 226, 233, 255, 211, 139, 59,
    118, 236, 245, 199, 163, 107, 214, 129, 47,  94,  188, 85,  170, 121,
    242, 201, 191, 83,  166, 97,  194, 169, 127, 254, 209, 143, 51,  102,
    204, 181, 71,  142, 49,  98,  196, 165, 103, 206, 177, 79,  158, 17,
    34,  68,  136, 61,  122, 244, 197, 167, 99,  198, 161, 111, 222, 145,
    15,  30,  60,  120, 240, 205, 183, 67,  134, 33,  66,  132, 37,  74,
    148, 5,   10,  20,  40,  80,  160, 109, 218, 153, 31,  62,  124, 248,
    221, 151, 3,   6,   12,  24,  48,  96,  192, 173, 119, 238, 241, 207,
    179, 75,  150
);

=head2 GFL

Log values used in the multiplication step of the Reed-Solomon encoding
algorithm.

=cut

our @GFL = (
    0,   0,   1,   240, 2,   225, 241, 53,  3,   38,  226, 133, 242, 43,
    54,  210, 4,   195, 39,  114, 227, 106, 134, 28,  243, 140, 44,  23,
    55,  118, 211, 234, 5,   219, 196, 96,  40,  222, 115, 103, 228, 78,
    107, 125, 135, 8,   29,  162, 244, 186, 141, 180, 45,  99,  24,  49,
    56,  13,  119, 153, 212, 199, 235, 91,  6,   76,  220, 217, 197, 11,
    97,  184, 41,  36,  223, 253, 116, 138, 104, 193, 229, 86,  79,  171,
    108, 165, 126, 145, 136, 34,  9,   74,  30,  32,  163, 84,  245, 173,
    187, 204, 142, 81,  181, 190, 46,  88,  100, 159, 25,  231, 50,  207,
    57,  147, 14,  67,  120, 128, 154, 248, 213, 167, 200, 63,  236, 110,
    92,  176, 7,   161, 77,  124, 221, 102, 218, 95,  198, 90,  12,  152,
    98,  48,  185, 179, 42,  209, 37,  132, 224, 52,  254, 239, 117, 233,
    139, 22,  105, 27,  194, 113, 230, 206, 87,  158, 80,  189, 172, 203,
    109, 175, 166, 62,  127, 247, 146, 66,  137, 192, 35,  252, 10,  183,
    75,  216, 31,  83,  33,  73,  164, 144, 85,  170, 246, 65,  174, 61,
    188, 202, 205, 157, 143, 169, 82,  72,  182, 215, 191, 251, 47,  178,
    89,  151, 101, 94,  160, 123, 26,  112, 232, 21,  51,  238, 208, 131,
    58,  69,  148, 18,  15,  16,  68,  17,  121, 149, 129, 19,  155, 59,
    249, 70,  214, 250, 168, 71,  201, 156, 64,  60,  237, 130, 111, 20,
    93,  122, 177, 150
);

=head2 POLY

Coefficients of the Reed-Solomon coeeficients.  The keys of the C<POLY> hash
correspond to the order of the respective polynomial.

=cut

our %POLY = (
    5  => [ 62,  111, 15,  48,  228 ],
    7  => [ 254, 92,  240, 134, 144, 68, 23 ],
    10 => [ 61,  110, 255, 116, 248, 223, 166, 185, 24, 28 ],
    11 => [ 120, 97,  60,  245, 39, 168, 194, 12, 205, 138, 175 ],
    12 => [ 242, 100, 178, 97,  213, 142, 42, 61, 91, 158, 153, 41 ],
    14 => [ 185, 83,  186, 18,  45, 138, 119, 157, 9, 95, 252, 192, 97, 156 ],
    18 => [
        188, 90, 48, 225, 254, 94,  129, 109, 213, 241,
        61,  66, 75, 188, 39,  100, 195, 83
    ],
    20 => [
        172, 186, 174, 27,  82, 108, 79, 253, 145, 153,
        160, 188, 2,   168, 71, 233, 9,  244, 195, 15
    ],
    24 => [
        193, 50,  96,  184, 181, 12,  124, 254, 172, 5,  21,  155,
        223, 251, 197, 155, 21,  176, 39,  109, 205, 88, 190, 52
    ],
    28 => [
        255, 93, 168, 233, 151, 120, 136, 141, 213, 110, 138, 17, 121, 249,
        34,  75, 53,  170, 151, 37,  174, 103, 96,  71,  97,  43, 231, 211
    ],
    36 => [
        112, 81,  98,  225, 25,  59,  184, 175, 44,  115, 119, 95,
        137, 101, 33,  68,  4,   2,   18,  229, 182, 80,  251, 220,
        179, 84,  120, 102, 181, 162, 250, 130, 218, 242, 127, 245
    ],
    42 => [
        5,   9,   5,   226, 177, 150, 50,  69,  202, 248, 101, 54,  57,  253,
        1,   21,  121, 57,  111, 214, 105, 167, 9,   100, 95,  175, 8,   242,
        133, 245, 2,   122, 105, 247, 153, 22,  38,  19,  31,  137, 193, 77
    ],
    48 => [
        19,  225, 253, 92,  213, 69,  175, 160, 147, 187, 87,  176,
        44,  82,  240, 186, 138, 66,  100, 120, 88,  131, 205, 170,
        90,  37,  23,  118, 147, 16,  106, 191, 87,  237, 188, 205,
        231, 238, 133, 238, 22,  117, 32,  96,  223, 172, 132, 245
    ],
    56 => [
        46,  143, 53,  233, 107, 203, 43,  155, 28,  247, 67,  127, 245, 137,
        13,  164, 207, 62,  117, 201, 150, 22,  238, 144, 232, 29,  203, 117,
        234, 218, 146, 228, 54,  132, 200, 38,  223, 36,  159, 150, 235, 215,
        192, 230, 170, 175, 29,  100, 208, 220, 17,  12,  238, 223, 9,   175
    ],
    62 => [
        204, 11,  47,  86,  124, 224, 166, 94,  7,   232, 107, 4,   170, 176,
        31,  163, 17,  188, 130, 40,  10,  87,  63,  51,  218, 27,  6,   147,
        44,  161, 71,  114, 64,  175, 221, 185, 106, 250, 190, 197, 63,  245,
        230, 134, 112, 185, 37,  196, 108, 143, 189, 201, 188, 202, 118, 39,
        210, 144, 50,  169, 93,  242
    ],
    68 => [
        186, 82,  103, 96,  63,  132, 153, 108, 54,  64,  189, 211, 232, 49,
        25,  172, 52,  59,  241, 181, 239, 223, 136, 231, 210, 96,  232, 220,
        25,  179, 167, 202, 185, 153, 139, 66,  236, 227, 160, 15,  213, 93,
        122, 68,  177, 158, 197, 234, 180, 248, 136, 213, 127, 73,  36,  154,
        244, 147, 33,  89,  56,  159, 149, 251, 89,  173, 228, 220
    ],
);

our @FORMATS = (
    [ 10,  10,  8,  8,  1,  8,   8,   3,    5,   3,   5,  1 ],
    [ 12,  12,  10, 10, 1,  10,  10,  5,    7,   5,   7,  1 ],
    [ 14,  14,  12, 12, 1,  12,  12,  8,    10,  8,   10, 1 ],
    [ 16,  16,  14, 14, 1,  14,  14,  12,   12,  12,  12, 1 ],
    [ 18,  18,  16, 16, 1,  16,  16,  18,   14,  18,  14, 1 ],
    [ 20,  20,  18, 18, 1,  18,  18,  22,   18,  22,  18, 1 ],
    [ 22,  22,  20, 20, 1,  20,  20,  30,   20,  30,  20, 1 ],
    [ 24,  24,  22, 22, 1,  22,  22,  36,   24,  36,  24, 1 ],
    [ 26,  26,  24, 24, 1,  24,  24,  44,   28,  44,  28, 1 ],
    [ 32,  32,  14, 14, 4,  28,  28,  62,   36,  62,  36, 1 ],
    [ 36,  36,  16, 16, 4,  32,  32,  86,   42,  86,  42, 1 ],
    [ 40,  40,  18, 18, 4,  36,  36,  114,  48,  114, 48, 1 ],
    [ 44,  44,  20, 20, 4,  40,  40,  144,  56,  144, 56, 1 ],
    [ 48,  48,  22, 22, 4,  44,  44,  174,  68,  174, 68, 1 ],
    [ 52,  52,  24, 24, 4,  48,  48,  204,  84,  102, 42, 2 ],
    [ 64,  64,  14, 14, 16, 56,  56,  280,  112, 140, 56, 2 ],
    [ 72,  72,  16, 16, 16, 64,  64,  368,  144, 92,  36, 4 ],
    [ 80,  80,  18, 18, 16, 72,  72,  456,  192, 114, 48, 4 ],
    [ 88,  88,  20, 20, 16, 80,  80,  576,  224, 144, 56, 4 ],
    [ 96,  96,  22, 22, 16, 88,  88,  696,  272, 174, 68, 4 ],
    [ 104, 104, 24, 24, 16, 96,  96,  816,  336, 136, 56, 6 ],
    [ 120, 120, 18, 18, 36, 108, 108, 1050, 496, 175, 68, 6 ],
    [ 132, 132, 20, 20, 36, 120, 120, 1304, 496, 163, 62, 8 ],
    [ 144, 144, 22, 22, 36, 132, 132, 1558, 620, 156, 62, 10 ],
    [ 8,   18,  6,  16, 1,  6,   16,  5,    7,   5,   7,  1 ],
    [ 8,   32,  6,  14, 2,  6,   28,  10,   11,  10,  11, 1 ],
    [ 12,  26,  10, 24, 1,  10,  24,  16,   14,  16,  14, 1 ],
    [ 12,  36,  10, 16, 2,  10,  32,  22,   18,  22,  18, 1 ],
    [ 16,  36,  14, 16, 2,  14,  32,  32,   24,  32,  24, 1 ],
    [ 16,  48,  14, 22, 2,  14,  44,  49,   28,  49,  28, 1 ],
);

our @C1 = (
    [ 0, 0 ],
    [ 0, 1 ],
    [ 0, 2 ],
    [ 0, 3 ],
    [ 0, 4 ],
    [ 0, 5 ],
    [ 0, 6 ],
    [ 0, 7 ],
    [ 0, 8 ],
    [ 0, 9 ],
    [ 0, 10 ],
    [ 0, 11 ],
    [ 0, 12 ],
    [ 0, 13 ],
    [ 0, 14 ],
    [ 0, 15 ],
    [ 0, 16 ],
    [ 0, 17 ],
    [ 0, 18 ],
    [ 0, 19 ],
    [ 0, 20 ],
    [ 0, 21 ],
    [ 0, 22 ],
    [ 0, 23 ],
    [ 0, 24 ],
    [ 0, 25 ],
    [ 0, 26 ],
    [ 0, 27 ],
    [ 0, 28 ],
    [ 0, 29 ],
    [ 0, 30 ],
    [ 0, 31 ],
    [3],
    [ 1, 0 ],
    [ 1, 1 ],
    [ 1, 2 ],
    [ 1, 3 ],
    [ 1, 4 ],
    [ 1, 5 ],
    [ 1, 6 ],
    [ 1, 7 ],
    [ 1, 8 ],
    [ 1, 9 ],
    [ 1, 10 ],
    [ 1, 11 ],
    [ 1, 12 ],
    [ 1, 13 ],
    [ 1, 14 ],
    [4],
    [5],
    [6],
    [7],
    [8],
    [9],
    [10],
    [11],
    [12],
    [13],
    [ 1, 15 ],
    [ 1, 16 ],
    [ 1, 17 ],
    [ 1, 18 ],
    [ 1, 19 ],
    [ 1, 20 ],
    [ 1, 21 ],
    [14],
    [15],
    [16],
    [17],
    [18],
    [19],
    [20],
    [21],
    [22],
    [23],
    [24],
    [25],
    [26],
    [27],
    [28],
    [29],
    [30],
    [31],
    [32],
    [33],
    [34],
    [35],
    [36],
    [37],
    [38],
    [39],
    [ 1, 22 ],
    [ 1, 23 ],
    [ 1, 24 ],
    [ 1, 25 ],
    [ 1, 26 ],
    [ 2, 0 ],
    [ 2, 1 ],
    [ 2, 2 ],
    [ 2, 3 ],
    [ 2, 4 ],
    [ 2, 5 ],
    [ 2, 6 ],
    [ 2, 7 ],
    [ 2, 8 ],
    [ 2, 9 ],
    [ 2, 10 ],
    [ 2, 11 ],
    [ 2, 12 ],
    [ 2, 13 ],
    [ 2, 14 ],
    [ 2, 15 ],
    [ 2, 16 ],
    [ 2, 17 ],
    [ 2, 18 ],
    [ 2, 19 ],
    [ 2, 20 ],
    [ 2, 21 ],
    [ 2, 22 ],
    [ 2, 23 ],
    [ 2, 24 ],
    [ 2, 25 ],
    [ 2, 26 ],
    [ 2, 27 ],
    [ 2, 28 ],
    [ 2, 29 ],
    [ 2, 30 ],
    [ 2, 31 ],
    [ 1, 30, 0, 0 ],
    [ 1, 30, 0, 1 ],
    [ 1, 30, 0, 2 ],
    [ 1, 30, 0, 3 ],
    [ 1, 30, 0, 4 ],
    [ 1, 30, 0, 5 ],
    [ 1, 30, 0, 6 ],
    [ 1, 30, 0, 7 ],
    [ 1, 30, 0, 8 ],
    [ 1, 30, 0, 9 ],
    [ 1, 30, 0, 10 ],
    [ 1, 30, 0, 11 ],
    [ 1, 30, 0, 12 ],
    [ 1, 30, 0, 13 ],
    [ 1, 30, 0, 14 ],
    [ 1, 30, 0, 15 ],
    [ 1, 30, 0, 16 ],
    [ 1, 30, 0, 17 ],
    [ 1, 30, 0, 18 ],
    [ 1, 30, 0, 19 ],
    [ 1, 30, 0, 20 ],
    [ 1, 30, 0, 21 ],
    [ 1, 30, 0, 22 ],
    [ 1, 30, 0, 23 ],
    [ 1, 30, 0, 24 ],
    [ 1, 30, 0, 25 ],
    [ 1, 30, 0, 26 ],
    [ 1, 30, 0, 27 ],
    [ 1, 30, 0, 28 ],
    [ 1, 30, 0, 29 ],
    [ 1, 30, 0, 30 ],
    [ 1, 30, 0, 31 ],
    [ 1, 30, 3 ],
    [ 1, 30, 1, 0 ],
    [ 1, 30, 1, 1 ],
    [ 1, 30, 1, 2 ],
    [ 1, 30, 1, 3 ],
    [ 1, 30, 1, 4 ],
    [ 1, 30, 1, 5 ],
    [ 1, 30, 1, 6 ],
    [ 1, 30, 1, 7 ],
    [ 1, 30, 1, 8 ],
    [ 1, 30, 1, 9 ],
    [ 1, 30, 1, 10 ],
    [ 1, 30, 1, 11 ],
    [ 1, 30, 1, 12 ],
    [ 1, 30, 1, 13 ],
    [ 1, 30, 1, 14 ],
    [ 1, 30, 4 ],
    [ 1, 30, 5 ],
    [ 1, 30, 6 ],
    [ 1, 30, 7 ],
    [ 1, 30, 8 ],
    [ 1, 30, 9 ],
    [ 1, 30, 10 ],
    [ 1, 30, 11 ],
    [ 1, 30, 12 ],
    [ 1, 30, 13 ],
    [ 1, 30, 1, 15 ],
    [ 1, 30, 1, 16 ],
    [ 1, 30, 1, 17 ],
    [ 1, 30, 1, 18 ],
    [ 1, 30, 1, 19 ],
    [ 1, 30, 1, 20 ],
    [ 1, 30, 1, 21 ],
    [ 1, 30, 14 ],
    [ 1, 30, 15 ],
    [ 1, 30, 16 ],
    [ 1, 30, 17 ],
    [ 1, 30, 18 ],
    [ 1, 30, 19 ],
    [ 1, 30, 20 ],
    [ 1, 30, 21 ],
    [ 1, 30, 22 ],
    [ 1, 30, 23 ],
    [ 1, 30, 24 ],
    [ 1, 30, 25 ],
    [ 1, 30, 26 ],
    [ 1, 30, 27 ],
    [ 1, 30, 28 ],
    [ 1, 30, 29 ],
    [ 1, 30, 30 ],
    [ 1, 30, 31 ],
    [ 1, 30, 32 ],
    [ 1, 30, 33 ],
    [ 1, 30, 34 ],
    [ 1, 30, 35 ],
    [ 1, 30, 36 ],
    [ 1, 30, 37 ],
    [ 1, 30, 38 ],
    [ 1, 30, 39 ],
    [ 1, 30, 1, 22 ],
    [ 1, 30, 1, 23 ],
    [ 1, 30, 1, 24 ],
    [ 1, 30, 1, 25 ],
    [ 1, 30, 1, 26 ],
    [ 1, 30, 2, 0 ],
    [ 1, 30, 2, 1 ],
    [ 1, 30, 2, 2 ],
    [ 1, 30, 2, 3 ],
    [ 1, 30, 2, 4 ],
    [ 1, 30, 2, 5 ],
    [ 1, 30, 2, 6 ],
    [ 1, 30, 2, 7 ],
    [ 1, 30, 2, 8 ],
    [ 1, 30, 2, 9 ],
    [ 1, 30, 2, 10 ],
    [ 1, 30, 2, 11 ],
    [ 1, 30, 2, 12 ],
    [ 1, 30, 2, 13 ],
    [ 1, 30, 2, 14 ],
    [ 1, 30, 2, 15 ],
    [ 1, 30, 2, 16 ],
    [ 1, 30, 2, 17 ],
    [ 1, 30, 2, 18 ],
    [ 1, 30, 2, 19 ],
    [ 1, 30, 2, 20 ],
    [ 1, 30, 2, 21 ],
    [ 1, 30, 2, 22 ],
    [ 1, 30, 2, 23 ],
    [ 1, 30, 2, 24 ],
    [ 1, 30, 2, 25 ],
    [ 1, 30, 2, 26 ],
    [ 1, 30, 2, 27 ],
    [ 1, 30, 2, 28 ],
    [ 1, 30, 2, 29 ],
    [ 1, 30, 2, 30 ],
    [ 1, 30, 2, 31 ],
);

1;
